Add 3rd dim, display trails and spawn attractor/repulsor on click

tristandeborde opened the pull request:
Hi,

I stumbled onto this project from your website, and I know this is very anecdotal to the FPGA magic but I was rly curious to know what this would look like in 3d so here we are... 

https://github.com/alienflip/N-Body-Problem/assets/33256624/b0d10124-6050-4ab4-8c0d-298c13770b08

A few notes:
- As you can see in the video, left mouse button spawns an "attractor" with positive mass; right click button spawns a "repulsor" with negative mass. Its position is fixed in the depth dimension cuz we place it with the cursor which is a 2d object ofc. 
- btw I have 0 physics knowledge so mistakes must have been made, pls tell me if you spot some !
- In terms of UX, aside from the bodies' size being rendered with spheres instead of points, it' can be tough to understand that there's a 3rd dimension. Maybe rotating the camera around the center of the scene would improve the user's 3d feeling ? I tried to do it quickly but did not have the time to make it work.
- Would you potentially be okay with me doing a three.js port of this to display it on a website (a bit like you did on yours?)

Cheers

alienflip commented:
This is epic!

Absolutely; feel free to add it to your site 🙂

In terms of the physics - I will take a look and then we can merge it...Could you provide a little explanation of your thought process this aspect of the pull request? 🙏

tristandeborde commented:
Thanks for the quick answer! Great to hear that you liked this :) 

I've added a readme to the Python dir, with the video as you suggested, pls tell me if you want some things changed.
As for the thought process, I'm not really sure what you're referring to, so I've added a feature list in the readme to better understand the 3d version's raison d'etre.

Among these is one feature that I should have thought about before: collision detection and body merging ofc!

Will implement this when I have the time in the coming days - we can discuss the details beforehand if you want. 
Btw, u should check your YC emails as I messaged you on there 😅 (that's how I stumbled onto this project!)

alienflip commented:
Awesome looks good; merged 🥶
